src/Control/Effect/Choose.hs

```diff
@@ -0,0 +1,95 @@
+{-# LANGUAGE DeriveGeneric, DeriveTraversable, FlexibleContexts, FlexibleInstances, MultiParamTypeClasses, RankNTypes, TypeOperators, UndecidableInstances #-}
+module Control.Effect.Choose
+( -- * Choose effect
+  Choose(..)
+, choose
+  -- * Choose carrier
+, runChoose
+, ChooseC(..)
+) where
+
+import Control.Applicative ((<|>), liftA2)
+import Control.Effect.Carrier
+import Control.Monad (join)
+import qualified Control.Monad.Fail as Fail
+import Control.Monad.Fix
+import Control.Monad.IO.Class
+import Control.Monad.Trans.Class
+import Data.Bool (bool)
+import Data.Maybe (fromJust)
+import GHC.Generics (Generic1)
+
+data Choose m k
+  = Choose (Bool -> m k)
+  deriving (Functor, Generic1)
+
+instance HFunctor Choose
+instance Effect   Choose
+
+-- | Nondeterministically choose between two computations.
+choose :: (Carrier sig m, Member Choose sig) => m a -> m a -> m a
+choose a b = send (Choose (bool b a))
+
```

Comment:
We could consider defining functionality like `optional` against this interface, since it doesn’t require `empty`. Likewise, possibly something like `Data.List.NonEmpty.some1`.

Reply:
`some1` would actually require `Empty` as well! Mea culpa.

Reply:
no, wait

Reply:
Done! See below.